Objects/stringlib/fastsearch.h

```diff
@@ -324,222 +359,402 @@ typedef struct STRINGLIB(_pre) {
+    Py_ssize_t shift;
+    STRINGLIB_CHAR s_last;
+    // Counters
+    Py_ssize_t count = 0;
+    Py_ssize_t iloop = 0, ihits_last = 0;
+    Py_ssize_t ihits = 0, iloop_last = 0;
+    // Loop
+    for (i = 0; i <= w;) {
+        iloop++;
+        ip = reversed ? -i : i;
+        s_last = ss[ip];
+        LOG2("Last window ch: %c\n", s_last);
+        if (true_gap) {
+            shift = 0;
+            if (s_last != p_last) {
+                if (i < w && !STRINGLIB_BLOOM(mask, ss[ip+dir])) {
+                    shift = m_p1;
+                }
+                else {
+                    shift = Py_MAX(table[s_last & TABLE_MASK], 1);
+                }
+            }
+        }
+        else {
+            shift = table[s_last & TABLE_MASK];
+        }
+        if (shift) {
+            LOG("Shift: %ld\n", shift);
+            i += shift;
+            continue;
+        }
+        // assert(s_last == p_last);                               // true_gap
+        // assert((s_last & TABLE_MASK) == (p_last & TABLE_MASK)); // else
+        j_off = ip - p_end;
+        for (j = 0; j < j_stop; j++) {
+            ihits++;
+            jp = p_stt + (reversed ? -j : j);
+            LOG2("Checking j=%ld: %c vs %c\n", j, ss[j_off + jp], p[jp]);
+            if (ss[j_off + jp] != p[jp]) {
+                break;
+            }
+        }
+        if (j == j_stop) {
+            LOG("Found a match!\n");
+            if (mode != FAST_COUNT) {
+                return reversed ? n - m - i : i;
+            }
+            if (++count == maxcount) {
+                return maxcount;
+            }
+            i += m;
+        }
+        else if (true_gap) {
+            if (i < w && !STRINGLIB_BLOOM(mask, ss[ip+dir])) {
+                LOG("Move by (m + 1) = %ld\n", m_p1);
+                i += m_p1;
+            }
+            else {
+                LOG("Move by true gap = %ld\n", gap);
+                i += true_gap;
+            }
+        }
+        else {
+            LOG("Move by table gap = %ld\n", gap);
+            i += gap;
+        }
+        if (dynamic) {
+            if (ihits - ihits_last < 100 && iloop - iloop_last < 100) {
+                continue;
+            }
+            ll = (double)(w - i + 1);
+            exp_hrs = ((double)iloop * hrs_lcost +
+                       (double)ihits * hrs_hcost) / (double)i * ll;
+            exp_twy = twy_icost + ll * twy_lcost;
+            if (exp_twy < exp_hrs) {
+                LOG("switching to two-way algorithm: n=%ld, m=%ld\n", n, m);
+                STRINGLIB(_init_critical_fac)(&pw, dir);
+                Py_ssize_t res = STRINGLIB(_two_way)(
+                        reversed ? s : s + i, n - i,
+                        maxcount - count, mode, &pw, dir);
+                if (mode == FAST_SEARCH) {
+                    return res == -1 ? -1 : (reversed ? res : res + i);
+                }
+                else {
+                    return res + count;
+                }
```

Comment:
I don't think that is an improvement.

Reply:
In light of your previous comment, it's not. But I personally find the if/else construct with returns inside too redundant (especially when there are already a lot of nesting in the code) but since it's not PEP 7 compliant, just forget about it.

Reply:
and this comment